apis/Google.Cloud.EntityFrameworkCore.Spanner/Google.Cloud.EntityFrameworkCore.Spanner/Storage/Internal/SpannerRelationalCommandBuilderFactory.cs

```diff
@@ -0,0 +1,374 @@
+        protected override IRelationalCommandBuilder CreateCore(
+            IDiagnosticsLogger<DbLoggerCategory.Database.Command> logger,
+            IRelationalTypeMapper relationalTypeMapper)
+            => new SpannerRelationalCommandBuilder(logger, relationalTypeMapper);
+
+        private sealed class SpannerRelationalCommandBuilder : RelationalCommandBuilder
+        {
+            public SpannerRelationalCommandBuilder(
+                IDiagnosticsLogger<DbLoggerCategory.Database.Command> logger,
+                IRelationalTypeMapper typeMapper)
+                : base(logger, typeMapper)
+            {
+            }
+
+            protected override IRelationalCommand BuildCore(
+                IDiagnosticsLogger<DbLoggerCategory.Database.Command> logger,
+                string commandText,
+                IReadOnlyList<IRelationalParameter> parameters)
+                => new SpannerRelationalCommand(logger, commandText, parameters);
+
+            private sealed class SpannerRelationalCommand : RelationalCommand
+            {
+                public SpannerRelationalCommand(
+                    IDiagnosticsLogger<DbLoggerCategory.Database.Command> logger,
+                    string commandText,
+                    IReadOnlyList<IRelationalParameter> parameters)
+                    : base(
+                        logger,
+                        commandText,
+                        parameters)
+                {
+                }
+
+                protected override object Execute(
+                    IRelationalConnection connection,
+                    DbCommandMethod executeMethod,
+                    IReadOnlyDictionary<string, object> parameterValues)
+                {
+                    GaxPreconditions.CheckNotNull(connection, nameof(connection));
+
+                    var dbCommand = CreateCommand(connection, parameterValues);
+
+                    connection.Open();
+
+                    var commandId = Guid.NewGuid();
+                    var startTime = DateTimeOffset.UtcNow;
+                    var stopwatch = Stopwatch.StartNew();
+
+                    Logger.CommandExecuting(
+                        dbCommand,
+                        executeMethod,
+                        commandId,
+                        connection.ConnectionId,
+                        false,
+                        startTime);
+
+                    object result;
+                    var readerOpen = false;
+
+                    try
+                    {
+                        //Note that adds/updates/deletes get intercepted at the 
+                        //modificationcommandbatch level and normally do not get translated
+                        //into ADO.NET commands here.  However, there are features in EF that
+                        //allow raw DML to be sent to the database which end up here and we need
+                        //to create the proper command and throw a useful exception.
```

Comment:
Yes.  Exactly.  I also have not completely decided whether to support this in some way.  eg, you can call ExecuteSqlCommand(string) to send a raw DML to the DB.  It'd be nice for us to provide some alternative direct access.  For example, it could be sending a configured command (ExecuteCommand(spannercommand)) which is a bit superflous -- or it could be a very simplified string syntax for insert/update/delete but that is likely not viable as we don't want to conflict if/when any other support comes from the backend to make this better.